Fix auto exponent issue for torch.pow

anjali411 opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/46936

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47024 Fix auto exponent issue for torch.pow**

Differential Revision: [D24698027](https://our.internmc.facebook.com/intern/diff/D24698027)

anjali411 committed:
Fix auto exponent issue for torch.pow

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"




[ghstack-poisoned]

ailzhang reviewed (approved):
Thanks!

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -185,8 +185,7 @@ Tensor norm_backward(Tensor grad, const Tensor & self, const optional<Scalar> &
 Tensor pow_backward(Tensor grad, const Tensor & self, const Scalar & exponent_) {
-  auto exponent = (exponent_.isComplex()) ? exponent_.toComplexDouble() : exponent_.toDouble();
-  if (exponent == 0.0) {
+  if ((exp.isComplex() && (exp.toComplexDouble() == 0.0)) || (!exp.isComplex() && (exp.toDouble() == 0.0))) {
```
I assume this is written this way because "base == 0.0" doesn't work. Out of curiosity, what is the technical reason for why "base == 0.0" doesn't work?

zou3519 replied:
Nvm, it looks like Scalar doesn't have any math operators implemented on it (which is probably the right call)

zou3519 commented on aten/src/ATen/native/Pow.cpp:
```diff
@@ -31,11 +31,13 @@ Tensor& pow_out(Tensor& result, const Tensor& base, Scalar exp) {
-
-  if (exponent == 0.0) {
+  if (exp.isComplex() && (exp.toComplexDouble() == 0.0) ) {
+    result.resize_as_(base).fill_(1);
+  } else if (exp.isComplex() && (exp.toComplexDouble() == 1.0) ) {
+    result.resize_as_(base).copy_(base);
+  } else if (!exp.isComplex() && (exp.toDouble() == 0.0)) {
     result.resize_as_(base).fill_(1);
-  } else if (exponent == 1.0) {
+  } else if (!exp.isComplex() && (exp.toDouble() == 1.0)) {
     result.resize_as_(base).copy_(base);
   } else {
```
Can this just be:
```
auto exponent = exp.toComplexDouble();
if (exponent == 0.0) {
...
} else if (exponent == 1.0) {
...
} else {
...
}
```?
`exponent` doesn't actually participate in the computation so it doesn't affect the type promotion here (but maybe we should leave a comment here because it's very easy to make a mistake here)

anjali411 replied:
this is resolved with the `equal` function now

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -185,8 +185,7 @@ Tensor norm_backward(Tensor grad, const Tensor & self, const optional<Scalar> &
 Tensor pow_backward(Tensor grad, const Tensor & self, const Scalar & exponent_) {
-  auto exponent = (exponent_.isComplex()) ? exponent_.toComplexDouble() : exponent_.toDouble();
-  if (exponent == 0.0) {
+  if ((exp.isComplex() && (exp.toComplexDouble() == 0.0)) || (!exp.isComplex() && (exp.toDouble() == 0.0))) {
```
I feel like there are going to be other places in the codebase where we want to compare a Scalar against a particular value. I'm not sure if there's a good reason why we don't have operator==.

Should we provide something like a helper function `is_scalar_equal(Scalar, double)`?

anjali411 replied:
makes sense! added an `equal` method

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -222,7 +221,12 @@ Tensor pow_backward_exponent(Tensor grad, const Tensor& self, const Tensor& expo
 
 Tensor pow_backward_exponent(Tensor grad, const Scalar & base, const Tensor& exponent, Tensor result) {
-  auto base_ = base.isComplex() ? base.toComplexDouble() : base.toDouble();
+  Scalar base_;
+  if (base.isComplex()) {
+    base_ = base.toComplexDouble();
+  } else {
+    base_ = base.toDouble();
+  }
```
This looks really fishy. `base_` is going to be the same Scalar as `base`; what the code is doing here is unwrap the Scalar and then rewrap it into a Scalar

anjali411 replied:
yeah updated

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47024 Fix auto exponent issue for torch.pow**

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47024 Fix auto exponent issue for torch.pow**

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47024 Fix auto exponent issue for torch.pow**

[ghstack-poisoned]

anjali411 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -222,7 +221,12 @@ Tensor pow_backward_exponent(Tensor grad, const Tensor& self, const Tensor& expo
 
 Tensor pow_backward_exponent(Tensor grad, const Scalar & base, const Tensor& exponent, Tensor result) {
-  auto base_ = base.isComplex() ? base.toComplexDouble() : base.toDouble();
+  Scalar base_;
```
yeah updated

anjali411 commented on c10/core/Scalar.h:
```diff
@@ -88,6 +88,17 @@ class C10_API Scalar {
   Scalar operator-() const;
 
+  template<typename T>
+  bool equal(T num) const {
```
cc. @ezyang to verify if it's ok to add `equal` for `Scalar`.

ezyang replied:
This seems fine.

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47024 Fix auto exponent issue for torch.pow**

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -222,9 +224,14 @@ Tensor pow_backward_exponent(Tensor grad, const Tensor& self, const Tensor& expo
+  auto grad_lambda = [](Tensor a, Scalar b) {
+    return AT_DISPATCH_DOUBLE_COMPLEXDOUBLE(b.type(), "scalar_val", ([&] {
+      scalar_t val = b.to<scalar_t>();
+      return (a * std::log(val)).conj();
```
This seems a little goofy. Why not also add a `log()` method on `Scalar`?

zou3519 replied:
Should Scalar have more operations defined on it? For example, in `pow_backward`, we would need to define an operator- for Scalar to avoid the dispatch

anjali411 replied:
@ezyang added log.
I think it makes sense to support more math operations for `Scalar`. For now, I added `log`, and `equal` in `Scalar.h` but we should add a new file `ScalarMath.{h,cpp}` if we plan to add more ops for scalar

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47024 Fix auto exponent issue for torch.pow**

[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -184,12 +184,12 @@ Tensor norm_backward(Tensor grad, const Tensor & self, const optional<Scalar> &
     return at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
   } else {
-    auto out = grad * (exponent * self.pow(exponent - 1)).conj();
+    auto grad_lambda = [&](auto exp) { return grad * (exp * self.pow(exp - 1)).conj(); };
```
Time to read up on auto type deduction...

zou3519 replied:
According to https://en.cppreference.com/w/cpp/language/lambda, `grad_lambda` is a generic lambda, which behaves like a template with one parameter. So this should work.

It's too bad that we can't actually write a test for this

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936


Differential Revision: [D24698027](https://our.internmc.facebook.com/intern/diff/D24698027)

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936


Differential Revision: [D24698027](https://our.internmc.facebook.com/intern/diff/D24698027)

[ghstack-poisoned]

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936


Differential Revision: [D24698027](https://our.internmc.facebook.com/intern/diff/D24698027)

[ghstack-poisoned]

albanD commented on test/cpp/api/autograd.cpp:
```diff
@@ -168,24 +168,26 @@ TEST(AutogradAPITests, AnomalyMode) {
+  //   auto y = x.pow(1.5);
+  //   auto gr =
+  //       grad({y}, {x}, {}, /*retain_graph=*/true, /*create_backward=*/true);
+  //   ASSERT_THROWS_WITH(grad({gr[0]}, {x});, "returned nan");
```
Can you try adding the grad_output as `torch::tensor({0.0})` here? That should make it `nan` as you expect.

anjali411 replied:
yup will do

anjali411 committed:
Update on "Fix auto exponent issue for torch.pow"

Fixes https://github.com/pytorch/pytorch/issues/46936


Differential Revision: [D24698027](https://our.internmc.facebook.com/intern/diff/D24698027)

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken pytorch_linux_xenial_py3_clang5_asan_test2. Relevant snippet:

```
Nov 15 08:59:54 /var/lib/jenkins/workspace/build/aten/src/ATen/native/cpu/PowKernel.cpp.AVX2.cpp:41:5: runtime error: division by zero
Nov 15 08:59:55     #0 0x7f18784fe544 in void at::native::(anonymous namespace)::basic_loop<at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const::{lambda()#2}::operator()() const::{lambda(float)#4}>(char* restrict*, long const*, long, long, at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const::{lambda()#2}::operator()() const::{lambda(float)#4}&&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x15241544)
Nov 15 08:59:55     #1 0x7f18784fdcad in void at::native::(anonymous namespace)::vectorized_loop<at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const::{lambda()#2}::operator()() const::{lambda(float)#4}, at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const::{lambda()#2}::operator()() const::{lambda(at::vec256::(anonymous namespace)::Vec256<float>)#4}>(char**, long, long, at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const::{lambda()#2}::operator()() const::{lambda(float)#4}&&, at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const::{lambda()#2}::operator()() const::{lambda(at::vec256::(anonymous namespace)::Vec256<float>)#4}&&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x15240cad)
Nov 15 08:59:55     #2 0x7f186f71de0e in void c10::function_ref<void (char**, long const*, long, long)>::callback_fn<at::TensorIterator::for_each(c10::function_ref<void (char**, long const*, long)>, long)::$_5>(long, char**, long const*, long, long) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xc460e0e)
Nov 15 08:59:55     #3 0x7f186f70d580 in at::TensorIterator::serial_for_each(c10::function_ref<void (char**, long const*, long, long)>, at::Range) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xc450580)
Nov 15 08:59:55     #4 0x7f186f70c71a in at::TensorIterator::for_each(c10::function_ref<void (char**, long const*, long, long)>, long) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xc44f71a)
Nov 15 08:59:55     #5 0x7f186f70c4b3 in at::TensorIterator::for_each(c10::function_ref<void (char**, long const*, long)>, long) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xc44f4b3)
Nov 15 08:59:55     #6 0x7f18784e3ee4 in at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar)::$_2::operator()() const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x15226ee4)
Nov 15 08:59:55     #7 0x7f18784dfce0 in at::native::(anonymous namespace)::pow_tensor_scalar_kernel(at::TensorIterator&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x15222ce0)
Nov 15 08:59:55     #8 0x7f186f29fb84 in void at::native::DispatchStub<void (*)(at::TensorIterator&, c10::Scalar), at::native::pow_tensor_scalar_stub>::operator()<at::TensorIterator&, c10::Scalar&>(c10::DeviceType, at::TensorIterator&, c10::Scalar&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xbfe2b84)
Nov 15 08:59:55     #9 0x7f186f29cc58 in at::native::pow_out(at::Tensor&, at::Tensor const&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xbfdfc58)
Nov 15 08:59:55     #10 0x7f186f29e461 in at::native::pow(at::Tensor const&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xbfe1461)
Nov 15 08:59:55     #11 0x7f18703e0b4a in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::Scalar), &at::(anonymous namespace)::pow_Tensor_Scalar>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::Scalar> >, at::Tensor (at::Tensor const&, c10::Scalar)>::call(c10::OperatorKernel*, at::Tensor const&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xd123b4a)
Nov 15 08:59:55     #12 0x7f186fd9c2fc in at::Tensor c10::KernelFunction::call<at::Tensor, at::Tensor const&, c10::Scalar>(c10::OperatorHandle const&, at::Tensor const&, c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xcadf2fc)
Nov 15 08:59:56     #13 0x7f186fd9b906 in at::Tensor c10::Dispatcher::callWithDispatchKey<at::Tensor, at::Tensor const&, c10::Scalar>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::Scalar)> const&, c10::DispatchKey, at::Tensor const&, c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xcade906)
Nov 15 08:59:56     #14 0x7f186fd9b4ef in at::Tensor c10::Dispatcher::call<at::Tensor, at::Tensor const&, c10::Scalar>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::Scalar)> const&, at::Tensor const&, c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xcade4ef)
Nov 15 08:59:56     #15 0x7f186fc0ce83 in at::pow(at::Tensor const&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xc94fe83)
Nov 15 08:59:56     #16 0x7f18749b9629 in torch::autograd::VariableType::(anonymous namespace)::pow_Tensor_Scalar(at::Tensor const&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x116fc629)
Nov 15 08:59:56     #17 0x7f18749b8f7a in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::Scalar), &torch::autograd::VariableType::(anonymous namespace)::pow_Tensor_Scalar>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::Scalar> >, at::Tensor (at::Tensor const&, c10::Scalar)>::call(c10::OperatorKernel*, at::Tensor const&, c10::Scalar) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x116fbf7a)
Nov 15 08:59:56     #18 0x7f186fd9c2fc in at::Tensor c10::KernelFunction::call<at::Tensor, at::Tensor const&, c10::Scalar>(c10::OperatorHandle const&, at::Tensor const&, c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xcadf2fc)
Nov 15 08:59:56     #19 0x7f186fd9b906 in at::Tensor c10::Dispatcher::callWithDispatchKey<at::Tensor, at::Tensor const&, c10::Scalar>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::Scalar)> const&, c10::DispatchKey, at::Tensor const&, c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xcade906)
Nov 15 08:59:56     #20 0x7f186fd9b4ef in at::Tensor c10::Dispatcher::call<at::Tensor, at::Tensor const&, c10::Scalar>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, c10::Scalar)> const&, at::Tensor const&, c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xcade4ef)
Nov 15 08:59:56     #21 0x7f18709773ca in at::Tensor::pow(c10::Scalar) const (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0xd6ba3ca)
Nov 15 08:59:56     #22 0x7f187643da1e in torch::autograd::generated::details::pow_backward(at::Tensor, at::Tensor const&, c10::Scalar const&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x13180a1e)
Nov 15 08:59:56     #23 0x7f18743c1267 in torch::autograd::generated::PowBackward0::apply(std::vector<at::Tensor, std::allocator<at::Tensor> >&&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x11104267)
Nov 15 08:59:56     #24 0x7f1875636bf8 in torch::autograd::Node::operator()(std::vector<at::Tensor, std::allocator<at::Tensor> >&&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x12379bf8)
Nov 15 08:59:56     #25 0x7f187561e83b in torch::autograd::Engine::evaluate_function(std::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&, std::shared_ptr<torch::autograd::ReadyQueue> const&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x1236183b)
Nov 15 08:59:56     #26 0x7f187561c86f in torch::autograd::Engine::thread_main(std::shared_ptr<torch::autograd::GraphTask> const&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x1235f86f)
Nov 15 08:59:56     #27 0x7f187562e527 in torch::autograd::Engine::execute_with_graph_task(std::shared_ptr<torch::autograd::GraphTask> const&, std::shared_ptr<torch::autograd::Node>) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x12371527)
Nov 15 08:59:56     #28 0x7f187562a134 in torch::autograd::Engine::execute(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, bool, bool, bool, std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x1236d134)
Nov 15 08:59:56     #29 0x7f187560ad6a in torch::autograd::run_backward(std::vector<at::Tensor, std::allocator<at::Tensor> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, bool, bool, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, bool, bool) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x1234dd6a)
Nov 15 08:59:56     #30 0x7f187560c0b1 in torch::autograd::grad(std::vector<at::Tensor, std::allocator<at::Tensor> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> > const&, c10::optional<bool>, bool, bool) (/var/lib/jenkins/workspace/build/lib/libtorch_cpu.so+0x1234f0b1)
Nov 15 08:59:56     #31 0x55651c46d554 in AutogradAPITests_AnomalyMode_Test::TestBody() (/var/lib/jenkins/workspace/build/bin/test_api+0xaf1554)
Nov 15 08:59:56     #32 0x55651d5e906a in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/var/lib/jenkins/workspace/build/bin/test_api+0x1c6d06a)
Nov 15 08:59:56     #33 0x55651d597d4a in testing::Test::Run() (/var/lib/jenkins/workspace/build/bin/test_api+0x1c1bd4a)
Nov 15 08:59:56     #34 0x55651d59a692 in testing::TestInfo::Run() (/var/lib/jenkins/workspace/build/bin/test_api+0x1c1e692)
Nov 15 08:59:56     #35 0x55651d59c5ba in testing::TestCase::Run() (/var/lib/jenkins/workspace/build/bin/test_api+0x1c205ba)
Nov 15 08:59:56     #36 0x55651d5be9c4 in testing::internal::UnitTestImpl::RunAllTests() (/var/lib/jenkins/workspace/build/bin/test_api+0x1c429c4)
Nov 15 08:59:56     #37 0x55651d5ee591 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (/var/lib/jenkins/workspace/build/bin/test_api+0x1c72591)
Nov 15 08:59:56     #38 0x55651d5bd469 in testing::UnitTest::Run() (/var/lib/jenkins/workspace/build/bin/test_api+0x1c41469)
Nov 15 08:59:56     #39 0x55651c45ada7 in main (/var/lib/jenkins/workspace/build/bin/test_api+0xadeda7)
Nov 15 08:59:56     #40 0x7f18594b683f in __libc_start_main /build/glibc-e6zv40/glibc-2.23/csu/../csu/libc-start.c:291
Nov 15 08:59:56     #41 0x55651c45a758 in _start (/var/lib/jenkins/workspace/build/bin/test_api+0xade758)
```
